### PR TITLE
2.x.x todos tutorial

### PR DESCRIPTION
**Pull request by adam-fowler**
This is the example used in the todos tutorial

### CODE REVIEW

File: todos-postgres-tutuorial/Sources/PostgresClientService.swift [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import Hummingbird
+@_spi(ConnectionPool) import PostgresNIO
```
Comment (Joannis) [inline todos-postgres-tutuorial/Sources/PostgresClientService.swift:2]: I like their connection pool, like everyone, but @_spi in a tutorial doesn't feel right

File: todos-postgres-tutuorial/Sources/TodoPostgresRepository.swift [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                logger: logger
+            )
+        }
+        return Todo(id: id, title: title, order: order, url: url, completed: nil)
+    }
+    /// Get todo.
+    func get(id: UUID) async throws -> Todo? { 
+        try await self.client.withConnection{ connection in
+            let stream = try await connection.query("""
+                SELECT "id", "title", "order", "url", "completed" FROM todos WHERE "id" = \(id)
```
Comment (Joannis) [inline todos-postgres-tutuorial/Sources/TodoPostgresRepository.swift:41]: Maybe worth noting that string-interpolation in safe in PostgresNIO?

File: todos-postgres-tutuorial/Sources/TodoPostgresRepository.swift [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            )
+        }
+    }
+    /// Create todo.
+    func create(title: String, order: Int?, urlPrefix: String) async throws -> Todo {
+        let id = UUID()
+        let url = urlPrefix + id.uuidString
+        _ = try await self.client.withConnection{ connection in 
+            try await connection.query(
+                "INSERT INTO todos (id, title, url, \"order\") VALUES (\(id), \(title), \(url), \(order));",
```
Comment (Joannis) [inline todos-postgres-tutuorial/Sources/TodoPostgresRepository.swift:31]: Maybe worth noting that string-interpolation in safe in PostgresNIO?

File: todos-postgres-tutuorial/Sources/TodoPostgresRepository.swift [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            }
+            return todos
+        }
+    }
+    /// Update todo. Returns updated todo if successful
+    func update(id: UUID, title: String?, order: Int?, completed: Bool?) async throws -> Todo? {
+        return try await self.client.withConnection{ connection in 
+            // UPDATE query
+            let query: PostgresQuery = """
+                UPDATE todos SET \(optionalUpdateFields: (("title", title), ("order", order), ("completed", completed))) WHERE id = \(id)
```
Comment (Joannis) [inline todos-postgres-tutuorial/Sources/TodoPostgresRepository.swift:70]: Note on interpolation

File: todos-postgres-tutuorial/Sources/TodoPostgresRepository.swift [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+                """
+            // if bind count is 1 then we aren't updating anything. Return nil
+            if query.binds.count == 1 {
+                return nil
+            }
+            _ = try await connection.query(query, logger: logger
+            )
+            // SELECT so I can get the full details of the TODO back
+            let stream = try await connection.query("""
+                SELECT "id", "title", "order", "url", "completed" FROM todos WHERE "id" = \(id)
```
Comment (Joannis) [inline todos-postgres-tutuorial/Sources/TodoPostgresRepository.swift:80]: Note on interpolation

File: todos-postgres-tutuorial/Sources/TodoPostgresRepository.swift [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+                return Todo(id: id, title: title, order: order, url: url, completed: completed)
+            }
+            return nil
+        }
+    }
+    /// Delete todo. Returns true if successful
+    func delete(id: UUID) async throws -> Bool {
+        return try await self.client.withConnection{ connection in
+            let selectStream = try await connection.query("""
+                SELECT "id" FROM todos WHERE "id" = \(id)
```
Comment (Joannis) [inline todos-postgres-tutuorial/Sources/TodoPostgresRepository.swift:93]: Note on interpolation

File: todos-postgres-tutuorial/Sources/Todos.swift [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+@main
+struct Todos: AsyncParsableCommand, AppArguments {
+    @Option(name: .shortAndLong)
+    var hostname: String = "127.0.0.1"
+
+    @Option(name: .shortAndLong)
+    var port: Int = 8080
+
+    var inMemoryTesting: Bool { false }
```
Comment (Joannis) [inline todos-postgres-tutuorial/Sources/Todos.swift:15]: Could this be a flag?

**Review by Joannis — CHANGES_REQUESTED** [review]

File: todos-postgres-tutuorial/Sources/PostgresClientService.swift (hunk 1 above)
Comment (adam-fowler) [reply todos-postgres-tutuorial/Sources/PostgresClientService.swift:2]: Yes I know. But I don't have an alternative unless I use the Vapor alternative. Remember this is still alpha software.

File: todos-postgres-tutuorial/Sources/TodoPostgresRepository.swift (hunk 2 above)
Comment (adam-fowler) [reply todos-postgres-tutuorial/Sources/TodoPostgresRepository.swift:41]: The actual tutorial text mentions this

File: todos-postgres-tutuorial/Sources/Todos.swift (hunk 7 above)
Comment (adam-fowler) [reply todos-postgres-tutuorial/Sources/Todos.swift:15]: Yep could be

File: todos-postgres-tutuorial/Sources/PostgresClientService.swift (hunk 1 above)
Comment (adam-fowler) [reply todos-postgres-tutuorial/Sources/PostgresClientService.swift:2]: I'll mention something in the tutorial text for the moment and hope Fabian removes this eventually

**Review by Joannis — APPROVED** [review]

**Pull request merged**